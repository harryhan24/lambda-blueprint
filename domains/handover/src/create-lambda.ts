import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";

export async function entrypoint(
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> {
  console.debug("Received event: %s", event);
  return { statusCode: 201, body: "" };
}
