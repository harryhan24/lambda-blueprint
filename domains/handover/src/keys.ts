export const enum Keys {
  TABLE_NAME = "TABLE_NAME",
  USERNAME = "USERNAME",
  PASSWORD = "PASSWORD"
}
