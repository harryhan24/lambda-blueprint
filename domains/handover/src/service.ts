import { CrudRepository, Error404 } from "@serverless-blueprint/core";
import { pipe } from "fp-ts/lib/pipeable"
import { fold } from "fp-ts/lib/Either"
import { v4 as uuidv4 } from "uuid";
import { CreateDto } from "./dtos/create-dto";
import { GetDto } from "./dtos/get-dto";
import { UpdateDto } from "./dtos/update-dto";
import { Handover } from "./entities/handover";

export class Service {
  constructor(
    private crudRepository: CrudRepository<Handover>
  ) {}

  async createHandover(createDto: CreateDto): Promise<void> {
    const date = new Date();

    const handover: Handover = {
      id: uuidv4(),
      createdAt: date.toISOString(),
      updatedAt: date.toISOString(),
      ...createDto,
    };
    return pipe(
      Handover.decode(handover),
      fold(
        // failure handler
        (reason) => Promise.reject(reason),
        // success handler
        (result) => this.crudRepository.put(result),
      ),
    );
    // Todo ...
  }

  async deleteHandover(id: string): Promise<void> {
    const keys: Partial<Handover> = { id: id };
    return this.crudRepository.delete(keys);
  }

  async getHandover(id: string): Promise<GetDto> {
    const keys: Partial<Handover> = { id: id };
    const item = await this.crudRepository.get(keys).catch((reason: any) => Promise.reject(reason));
    if (!item) { throw new Error404(); }
    const { createdAt, updatedAt, ...getDto } = item;
    return getDto;
  }

  async updateHandover(id: string, updateDto: UpdateDto): Promise<void> {}
}
