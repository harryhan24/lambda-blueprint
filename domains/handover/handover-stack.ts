import * as apigateway from "@aws-cdk/aws-apigateway";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as lambda from "@aws-cdk/aws-lambda";
import * as iam from "@aws-cdk/aws-iam";
import * as cdk from "@aws-cdk/core";
import { DomainStackProps } from "../../cdk/src/interfaces/domain-stack-props";
import { Keys } from "./src/utils/keys";

export class HandoverStack extends cdk.Stack {
  constructor(scope: cdk.App, props: DomainStackProps, path: string) {
    super(scope, `${props.env}-handover-stack`);

    const restApi = props.restApi;

    const dynamodbTable = new dynamodb.Table(this, "handovers", {
      tableName: `${props.env}-handovers`,
      partitionKey: {
        name: "id",
        type: dynamodb.AttributeType.STRING,
      },
    });

    const environment: Record<string, string> = {};

    environment[Keys.ENV]        = props.env;
    environment[Keys.TABLE_NAME] = dynamodbTable.tableName;
    // ...

    // Todo: build ssm-role and attach it ...

    const createLambda = new lambda.Function(this, "createLambda", {
      handler: "create-lambda-bundle.entrypoint",
      runtime: lambda.Runtime.NODEJS_12_X,
      code:    lambda.AssetCode.fromAsset(
        path,
        { exclude: ["**", "!create-lambda-bundle.js"] },
      ),
      environment,
    });

    const deleteLambda = new lambda.Function(this, "deleteLambda", {
      handler: "delete-lambda-bundle.entrypoint",
      runtime: lambda.Runtime.NODEJS_12_X,
      code:    lambda.AssetCode.fromAsset(
        path,
        { exclude: ["**", "!delete-lambda-bundle.js"] },
      ),
      environment,
    });

    const getLambda = new lambda.Function(this, "getLambda", {
      handler: "get-lambda-bundle.entrypoint",
      runtime: lambda.Runtime.NODEJS_12_X,
      code:    lambda.AssetCode.fromAsset(
        path,
        { exclude: ["**", "!get-lambda-bundle.js"] },
      ),
      environment,
    });

    const updateLambda = new lambda.Function(this, "updateLambda", {
      handler: "update-lambda-bundle.entrypoint",
      runtime: lambda.Runtime.NODEJS_12_X,
      code:    lambda.AssetCode.fromAsset(
        path,
        { exclude: ["**", "!update-lambda-bundle.js"] },
      ),
      environment,
    });

    dynamodbTable.grantWriteData(createLambda);
    dynamodbTable.grantWriteData(deleteLambda);
    dynamodbTable.grantReadData(getLambda);
    dynamodbTable.grantReadWriteData(updateLambda);

    const handovers = restApi.root.addResource("handovers");

    const createIntegration = new apigateway.LambdaIntegration(createLambda);
    handovers.addMethod("POST", createIntegration);

    const handover = handovers.addResource("{id}");

    const deleteIntegration = new apigateway.LambdaIntegration(deleteLambda);
    handover.addMethod("DELETE", deleteIntegration);

    const getIntegration = new apigateway.LambdaIntegration(getLambda);
    handover.addMethod("GET", getIntegration);

    const updateIntegration = new apigateway.LambdaIntegration(updateLambda);
    handover.addMethod("PUT", updateIntegration);

    handovers.addCorsPreflight({
      allowOrigins: apigateway.Cors.ALL_ORIGINS, // <--- Todo: Restrict?
      allowMethods: ["POST"],
    });
    handover.addCorsPreflight({
      allowOrigins: apigateway.Cors.ALL_ORIGINS, // <--- Todo: Restrict?
      allowMethods: ["DELETE, GET, PUT"],
    });
  }
}
