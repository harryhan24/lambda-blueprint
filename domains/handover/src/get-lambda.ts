import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { CrudRepository, Error404 } from "@serverless-blueprint/core";
import { Handover } from "./entities/handover";
import { Service } from "./service";

const tableName  = process.env.TABLE_NAME!; // Todo
const repository = new CrudRepository<Handover>({ tableName: tableName });
const service    = new Service(repository);

export async function entrypoint(
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> {
  console.debug("Received handover-event: %s", event);

  const pathParameters = event.pathParameters;
  if (pathParameters == null) { return { statusCode: 400, body: "" }; }

  try {
    const handoverDto = await service.getHandover(pathParameters["id"]);
    return { statusCode: 200, body: JSON.stringify(handoverDto) };
  } catch (reason) {
    console.debug(reason.constructor === Error404);
    return { statusCode: 500, body: "" };
    // Todo
  }
}
