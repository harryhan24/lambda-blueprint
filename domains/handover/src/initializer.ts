import { CrudRepository, decrypt } from "@serverless-blueprint/core";
import { Handover } from "./entities/handover";
import { Service } from "./service";
import { Keys } from "./keys";

export async function initialize(): Promise<Service> {
  // @ts-ignore
  process.env = await decrypt(process.env).catch((reason: any) => Promise.reject(reason));

  const tableName  = process.env[Keys.TABLE_NAME] || "";
  const repository = new CrudRepository<Handover>({ tableName: tableName });

  return new Service(repository);
}
