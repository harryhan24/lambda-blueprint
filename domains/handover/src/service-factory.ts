import { CrudRepository } from "@serverless-blueprint/core";
import { Handover } from "./handover";
import { Service } from "./service";
import { Keys } from "./keys";

export const createService = async (): Promise<Service> => {
  // Todo: ssm

  const tableName  = process.env[Keys.TABLE_NAME] || "";
  const repository = new CrudRepository<Handover>({ tableName: tableName });

  return new Service(repository);
};
