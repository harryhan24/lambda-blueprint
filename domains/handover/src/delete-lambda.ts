import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { handleError } from "@serverless-blueprint/core";
import { initialize } from "./initializer";
import { Service } from "./service";

let service: Service;

export async function entrypoint(
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> {
  console.debug("Received handover-event: %s", event);
  try {
    if (!service) {
      service = await initialize();
    }
    const pathParameters = event.pathParameters;
    if (pathParameters == null) { return { statusCode: 400, body: "" }; }

    await service.deleteHandover(pathParameters["id"]);
    return { statusCode: 204, body: "" };
  } catch (error) {
    return handleError(error);
    // Todo
  }
}
