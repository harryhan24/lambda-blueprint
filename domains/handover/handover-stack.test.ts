// https://docs.aws.amazon.com/cdk/latest/guide/testing.html

import * as cdk from "@aws-cdk/core";
import * as assert from "@aws-cdk/assert";

import { HandoverStack } from "./handover-stack";
import { ApigatewayStack } from "../../cdk/src/apigateway-stack";
import { SharedStackProps } from "../../cdk/src/interfaces/shared-stack-props";
import { DomainStackProps } from "../../cdk/src/interfaces/domain-stack-props";

describe("handover-stack", () => {
  it("test", async () => {
    const app = new cdk.App({});

    const sharedProps: SharedStackProps = { env: "local" };
    const apigatewayStack = new ApigatewayStack(app, sharedProps);

    const domainProps: DomainStackProps = { ...sharedProps, restApi: apigatewayStack.restApi }; // Todo: Mock?

    const handoverStack = new HandoverStack(app, domainProps, "dist");
    const cOutput = assert.SynthUtils.toCloudFormation(handoverStack);

    expect(cOutput).toMatchSnapshot();
  });
});
