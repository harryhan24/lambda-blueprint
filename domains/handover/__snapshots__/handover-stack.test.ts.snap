// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customer-stack test 1`] = `
Object {
  "Outputs": Object {
    "ExportsOutputFnGetAttcreateLambda327A7376Arn505586C9": Object {
      "Export": Object {
        "Name": "local-handover-stack:ExportsOutputFnGetAttcreateLambda327A7376Arn505586C9",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "createLambda327A7376",
          "Arn",
        ],
      },
    },
    "ExportsOutputFnGetAttdeleteLambda64E16E5CArn5C6E4DBF": Object {
      "Export": Object {
        "Name": "local-handover-stack:ExportsOutputFnGetAttdeleteLambda64E16E5CArn5C6E4DBF",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "deleteLambda64E16E5C",
          "Arn",
        ],
      },
    },
    "ExportsOutputFnGetAttgetLambda0E45CF34Arn12A2311A": Object {
      "Export": Object {
        "Name": "local-handover-stack:ExportsOutputFnGetAttgetLambda0E45CF34Arn12A2311A",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "getLambda0E45CF34",
          "Arn",
        ],
      },
    },
    "ExportsOutputFnGetAttupdateLambda2E646683ArnE6EC4AF9": Object {
      "Export": Object {
        "Name": "local-handover-stack:ExportsOutputFnGetAttupdateLambda2E646683ArnE6EC4AF9",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "updateLambda2E646683",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6ArtifactHashA3EBD759": Object {
      "Description": "Artifact hash for asset \\"612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6\\"",
      "Type": "String",
    },
    "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6S3BucketDEFFD7B1": Object {
      "Description": "S3 bucket for asset \\"612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6\\"",
      "Type": "String",
    },
    "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6S3VersionKey6DB2A398": Object {
      "Description": "S3 key for asset version \\"612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6\\"",
      "Type": "String",
    },
    "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813ArtifactHash89FA485F": Object {
      "Description": "Artifact hash for asset \\"cf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813\\"",
      "Type": "String",
    },
    "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813S3BucketB12E334E": Object {
      "Description": "S3 bucket for asset \\"cf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813\\"",
      "Type": "String",
    },
    "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813S3VersionKey02821A10": Object {
      "Description": "S3 key for asset version \\"cf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813\\"",
      "Type": "String",
    },
    "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293ArtifactHash44B44AC3": Object {
      "Description": "Artifact hash for asset \\"d4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293\\"",
      "Type": "String",
    },
    "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293S3Bucket034A9049": Object {
      "Description": "S3 bucket for asset \\"d4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293\\"",
      "Type": "String",
    },
    "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293S3VersionKey2BF60BDE": Object {
      "Description": "S3 key for asset version \\"d4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293\\"",
      "Type": "String",
    },
    "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837ArtifactHash59AF2C78": Object {
      "Description": "Artifact hash for asset \\"dc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837\\"",
      "Type": "String",
    },
    "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837S3Bucket4DA29CC5": Object {
      "Description": "S3 bucket for asset \\"dc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837\\"",
      "Type": "String",
    },
    "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837S3VersionKeyEBB50395": Object {
      "Description": "S3 key for asset version \\"dc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "createLambda327A7376": Object {
      "DependsOn": Array [
        "createLambdaServiceRoleDefaultPolicy7FCD0D30",
        "createLambdaServiceRole20E8740B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813S3BucketB12E334E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813S3VersionKey02821A10",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscf02d9b41920c94e6a1270201054b7f482d2d88785e4a9059e3d38fc26ad5813S3VersionKey02821A10",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "ENV": "local",
            "TABLE_NAME": Object {
              "Ref": "handovers6CB80F9B",
            },
          },
        },
        "Handler": "create-lambda-bundle.entrypoint",
        "Role": Object {
          "Fn::GetAtt": Array [
            "createLambdaServiceRole20E8740B",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "createLambdaServiceRole20E8740B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "createLambdaServiceRoleDefaultPolicy7FCD0D30": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "handovers6CB80F9B",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "createLambdaServiceRoleDefaultPolicy7FCD0D30",
        "Roles": Array [
          Object {
            "Ref": "createLambdaServiceRole20E8740B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "deleteLambda64E16E5C": Object {
      "DependsOn": Array [
        "deleteLambdaServiceRoleDefaultPolicy770F9D88",
        "deleteLambdaServiceRole44BC2A4A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837S3Bucket4DA29CC5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837S3VersionKeyEBB50395",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdc13e79edbff8e0cf44daf648d9ae7691c65090591adbdfaeec9730fdd65d837S3VersionKeyEBB50395",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "ENV": "local",
            "TABLE_NAME": Object {
              "Ref": "handovers6CB80F9B",
            },
          },
        },
        "Handler": "delete-lambda-bundle.entrypoint",
        "Role": Object {
          "Fn::GetAtt": Array [
            "deleteLambdaServiceRole44BC2A4A",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "deleteLambdaServiceRole44BC2A4A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deleteLambdaServiceRoleDefaultPolicy770F9D88": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "handovers6CB80F9B",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deleteLambdaServiceRoleDefaultPolicy770F9D88",
        "Roles": Array [
          Object {
            "Ref": "deleteLambdaServiceRole44BC2A4A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "getLambda0E45CF34": Object {
      "DependsOn": Array [
        "getLambdaServiceRoleDefaultPolicy76D4F95E",
        "getLambdaServiceRoleAF88CB09",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293S3Bucket034A9049",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293S3VersionKey2BF60BDE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd4c70a63a218d219921af3d7206314212f09094cf66f391ee7b5f51ec8ba2293S3VersionKey2BF60BDE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "ENV": "local",
            "TABLE_NAME": Object {
              "Ref": "handovers6CB80F9B",
            },
          },
        },
        "Handler": "get-lambda-bundle.entrypoint",
        "Role": Object {
          "Fn::GetAtt": Array [
            "getLambdaServiceRoleAF88CB09",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "getLambdaServiceRoleAF88CB09": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getLambdaServiceRoleDefaultPolicy76D4F95E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "handovers6CB80F9B",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getLambdaServiceRoleDefaultPolicy76D4F95E",
        "Roles": Array [
          Object {
            "Ref": "getLambdaServiceRoleAF88CB09",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "handovers6CB80F9B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "local-handovers",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "updateLambda2E646683": Object {
      "DependsOn": Array [
        "updateLambdaServiceRoleDefaultPolicyCD1BE1A0",
        "updateLambdaServiceRole660E4971",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6S3BucketDEFFD7B1",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6S3VersionKey6DB2A398",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters612c65cc95b62953969c00d19020163f7f2d45fec661ef1e672c10043e01bdb6S3VersionKey6DB2A398",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "ENV": "local",
            "TABLE_NAME": Object {
              "Ref": "handovers6CB80F9B",
            },
          },
        },
        "Handler": "update-lambda-bundle.entrypoint",
        "Role": Object {
          "Fn::GetAtt": Array [
            "updateLambdaServiceRole660E4971",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "updateLambdaServiceRole660E4971": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "updateLambdaServiceRoleDefaultPolicyCD1BE1A0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "handovers6CB80F9B",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "updateLambdaServiceRoleDefaultPolicyCD1BE1A0",
        "Roles": Array [
          Object {
            "Ref": "updateLambdaServiceRole660E4971",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
